src/intrinsics/ecma262/Object.js

```diff
@@ -101,36 +136,69 @@ export default function(realm: Realm): NativeFunctionValue {
         }
         keys = frm.$OwnPropertyKeys();
 
-        if (to_must_be_partial) {
-          if (to instanceof AbstractObjectValue && to.values.isTop()) {
-            // We don't know which objects to make partial and making all objects partial is failure in itself
-            AbstractValue.reportIntrospectionError(to);
-            throw new FatalError();
-          } else {
-            // if to has properties, we better remove them because after the temporal call to Object.assign we don't know their values anymore
-            if (to.hasStringOrSymbolProperties()) {
-              // preserve them in a snapshot and add the snapshot to the sources
-              delayedSources.push(to.getSnapshot({ removeProperties: true }));
-            }
+        handleSnapshot(frm, frm_was_partial);
+
+        // c. Repeat for each element nextKey of keys in List order,
+        invariant(frm, "from required");
+        invariant(keys, "keys required");
+        copyKeys(keys, frm, to);
+      };
+
+      const tryAndApplySourceOrRecover = nextSource => {
+        let effects;
+        let savedSuppressDiagnostics = realm.suppressDiagnostics;
+        try {
+          realm.suppressDiagnostics = true;
+          effects = realm.evaluateForEffects(
+            () => {
+              applySource(nextSource);
+              return realm.intrinsics.undefined;
+            },
+            undefined,
+            "tryAndApplySourceOrRecover"
+          );
+        } catch (e) {
+          // ignore null or undefined
+          if (nextSource === realm.intrinsics.null || nextSource === realm.intrinsics.undefined) {
+            return;
+          }
+          let frm = To.ToObject(realm, nextSource);
+
+          if (e instanceof FatalError && to.isSimpleObject()) {
```

Comment:
So technically, we should havoc only `frm` because it can have getters on it. @NTillmann and I have an idea for why we might be able to get away with not havocing on getters.

I actually think that this might not be as bad as it sounds. That will still let you do everything else in your tests here. 

However, I don't think havocing will be all that bad. If we knew anything about the properties here, then we wouldn't hit this branch it would just go through the normal application. If don't know anything about the properties, such as when we have a completely abstract object, then havocing doesn't really forget anything useful anyway.

Havocing might forget a bit more than necessary right now but we can fix that separately.

I'd prefer adding the havocing here in this first PR, for now and then remove it when we relax the requirements for getters later on. When we've narrowed down exactly what that means.

This should also be wrapped in realm.evaluateWithPossibleThrowCompletion since the getters might throw. I think we'll want that even if we end up not havocing the source.

Reply:
I really don't understand why you need to havoc the `frm` if it's a simple concrete object. It can't possibly have getters/setters on? What value does havocing it add here, maybe I'm missing something.

Reply:
Done.